### PR TITLE
Add tests for generate_encode.py and VideoWorker

### PR DESCRIPTION
**Pull request by mushtaqak**
This PR adds the tests to `__init__py` (VideoWorker)  and `generate_encode.py`

[EDUCATOR-920](https://openedx.atlassian.net/browse/EDUCATOR-920)

### CODE REVIEW

**Comment by muzaffaryousaf** [comment]
@mushtaqak FYI, Travis check ran for this PR.

File: video_worker/tests/test_video_worker.py [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+            """
+            Changes VM.VideoObject.valid to True.
+            """
+            self.VideoObject.valid = False
+
+        change_video_func = change_video_valid if is_valid else change_video_invalid
+        change_video_func_intake = change_video_valid_intake if is_valid_engine_intake else change_video_invalid_intake
+
+        with patch('video_worker.abstractions.Video.activate', new=change_video_func) as mock_video_activate, \
+                patch.object(VideoWorker, '_engine_intake', new=change_video_func_intake) as mock_engine_intake:
```
Comment (muhammad-ammar) [inline video_worker/tests/test_video_worker.py]: we can use nested context managers like

```python
with patch('video_worker.abstractions.Video.activate', new=change_video_func) as mock_video_activate:
    with patch.object(VideoWorker, '_engine_intake', new=change_video_func_intake) as mock_engine_intake:
```

File: video_worker/tests/test_video_worker.py [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+            Changes VM.VideoObject.valid to False.
+            """
+            self.VideoObject.valid = True
+
+        def change_video_invalid_intake(self):
+            """
+            Changes VM.VideoObject.valid to True.
+            """
+            self.VideoObject.valid = False
+
```
Comment (muhammad-ammar) [inline video_worker/tests/test_video_worker.py]: can we have 2 methods instead of 4 above?

File: video_worker/tests/test_video_worker.py [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+    @patch.object(VideoWorker, '_hls_pipeline')
+    @patch.object(VideoWorker, '_update_api')
+    @patch('video_worker.video_images.VideoImages.settings_setup')
+    @patch('video_worker.video_images.VideoImages.create_and_update')
+    @patch.object(deliverable_route, 'apply_async')
+    @patch.object(video_worker_logger, 'error')
+    @patch('shutil.rmtree')
+    @patch('os.path.exists')
+    @patch('os.mkdir')
+    def test_run(self, mock_data, mock_mkdir, mock_exists, mock_rmtree, mock_logger, celeryapp_sync_mock,
```
Comment (muhammad-ammar) [inline video_worker/tests/test_video_worker.py]: It is very hard to understand these big tests with so many patches. Can we make them simple?

File: video_worker/tests/test_video_worker.py [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+        mock_exists.return_value = False
+
+        self.assertIsNone(self.VW.endpoint_url)
+
+        self.VW._hls_pipeline()
+
+        self.assertIsNone(self.VW.endpoint_url)
+        mock_logger.assert_called_with('Source File (local) NOT FOUND - HLS')
+
+    @data(
```
Comment (muhammad-ammar) [inline video_worker/tests/test_video_worker.py]: why tuples with dicts? why not just dicts?

File: video_worker/tests/test_video_worker.py [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+
+        self.assertIsNone(self.VW.endpoint_url)
+
+        with patch('chunkey.Chunkey.__init__', new=change_chunkey) as mock_chunkey:
+            self.VW._hls_pipeline()
+
+        self.assertIsNotNone(self.VW.endpoint_url)
+        self.assertEqual(self.VW.endpoint_url, '/dummy-manifest-url')
+
+    @data(
```
Comment (muhammad-ammar) [inline video_worker/tests/test_video_worker.py]: same comment as above.

File: video_worker/tests/test_video_worker.py [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+        with patch('video_worker.abstractions.Encode.pull_data', new=pull_data_change_function) as pull_data_mock:
+            self.VW._generate_encode()
+
+        if filetype and ffcommand:
+            self.assertIsNotNone(self.VW.ffcommand)
+            self.assertEqual(self.VW.ffcommand, ffcommand)
+        else:
+            self.assertIsNone(self.VW.ffcommand)
+
+    @data(
```
Comment (muhammad-ammar) [inline video_worker/tests/test_video_worker.py]: same comment regarding dict inside tuple.

File: video_worker/tests/test_video_worker.py [hunk 7]
```diff
@@ -0,0 +1,535 @@
...
+            mock_logger.assert_called_with(mock_data.get('error_message', ''))
+
+        expected_output_file = 'ffcommand-outfile' if len(mock_data.get('path_exists', [])) else expected_output_file
+        self.assertEqual(self.VW.output_file, expected_output_file)
+
+    @data(True, False)
+    @patch('video_worker.validate.ValidateVideo.validate')
+    def test_validate_encode(self, is_valid, mock_valid):
+        """
+        Tests `_validate_encode` method workss correctly.
```
Comment (muhammad-ammar) [inline video_worker/tests/test_video_worker.py]: typo: works

File: video_worker/tests/test_generate_encode.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+        # Setup WS.settings_dict
+        WS = create_worker_setup()
+        WS.run()
+
+        with patch('video_worker.config.WorkerSetup.__init__', new=change_worker_setup) as ws_mock:
+            result_settings_dict = self.command_generate.settings_setup()
+            self.assertEqual(result_settings_dict, WS.settings_dict)
+
+    @data(
```
Comment (muhammad-ammar) [inline video_worker/tests/test_generate_encode.py]: why dicts inside tuple?

File: video_worker/tests/test_generate_encode.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+            self.assertEqual(command_generate.workdir, expected_workdir)
+            self.assertEqual(result_command, ' '.join(ffcommand))
+
+            if mock_data.get('ENFORCE_TARGET_ASPECT', False):
+                self.assertTrue(mock_scalar.called)
+
+            self.assertTrue(mock_call.called)
+            self.assertTrue(mock_codec.called)
+            self.assertTrue(mock_passes.called)
+            self.assertTrue(mock_destination.called)
```
Comment (muhammad-ammar) [inline video_worker/tests/test_generate_encode.py]: Everything is mocked here, can not we test the `generate` method without mocking. As per my understanding `generate` method is not doing any third party calls.

File: video_worker/generate_encode.py [hunk 10]
```diff
@@ -113,99 +98,116 @@ def _call(self):
...
+        # Append commands
+
+        # TODO-1:
+        # if mezz is None then next condition in or works wich would attempt to check
+        # float(None) technically which is bound to raise TypeError
+        # float() argument must be a string or a number, not 'NoneType'
+        # I believe it should be an `and` condition like below
+        # if mezz_aspect_ratio is not None and float(mezz_aspect_ratio) == float(TARGET_ASPECT_RATIO):
+
         if mezz_aspect_ratio is not None or float(mezz_aspect_ratio) == float(TARGET_ASPECT_RATIO):
```
Comment (mushtaqak) [inline video_worker/generate_encode.py]: @yro `mezz_aspect_ratio` would be generating `TypeError` if it is set to None. See comments above and below.

File: video_worker/generate_encode.py [hunk 11]
```diff
@@ -113,99 +98,116 @@ def _call(self):
...
 
-        if aspect_fix is False and resolution_fix is False:
-            return None
+        if not aspect_fix and not resolution_fix:
+            return
 
-        if aspect_fix is False and resolution_fix is True:
-            self.ffcommand.append("-vf")
-            self.ffcommand.append("scale=" + str(horiz_resolution) + ":" + str(self.EncodeObject.resolution))
+        ### MUSHI: COVERED UPTO THIS
```
Comment (mushtaqak) [inline video_worker/generate_encode.py]: typo - would prolly remove in next commit.

File: video_worker/generate_encode.py [hunk 12]
```diff
@@ -113,99 +98,116 @@ def _call(self):
...
 
-        elif aspect_fix is True:
+        if not aspect_fix and resolution_fix:
+            self.ffcommand.append('-vf')
+            self.ffcommand.append('scale=' + str(horiz_resolution) + ':' + str(self.EncodeObject.resolution))
+        # TODO-1: Flow would never reaches this line due to TypeError on mezz_aspect_ratio when it is None
```
Comment (mushtaqak) [inline video_worker/generate_encode.py]: @yro @muzaffaryousaf This is a concern. Though coverage would not work for this but I fear this block never executes. Can you guys suggest what should we do here ?

**Comment by mushtaqak** [comment]
@muzaffaryousaf @muhammad-ammar  @yro This is ready for review.

I have added coverage for `_scalar` of `generate_encode` as much as I can but it looks some block never executes so skipping for that.

File: video_worker/__init__.py [hunk 13]
```diff
@@ -126,36 +112,31 @@ def run(self):
...
-          II. change status in APIs
-          III. Generate Encode Command
-          IV. Execute Encodes
-            IVa. Validate Products
-          (*)V. Deliver Encodes (sftp and others?), retrieve URLs
-          (*)VI. Change Status in APIs, add URLs
-          VII. Clean Directory
+        # Pipeline Steps :
+        #   I. Intake
+        #     Ib. Validate Mezz
```
Comment (yro) [inline video_worker/__init__.py]: It's short for "mezzanine", it's a video term for an intermediate file (or source file in this case).

File: video_worker/generate_encode.py (hunk 10 above)
Comment (yro) [inline video_worker/generate_encode.py]: @mushtaqak Correct -- There are times where ffprobe gets weird metadata (from an exotic filetype for example, or if theres something unusual in the metadata), and this was intended as a protective measure against that. I'm happy to go with your recommendations on this.

File: video_worker/generate_encode.py (hunk 12 above)
Comment (yro) [inline video_worker/generate_encode.py]: @mushtaqak If mezz_aspect_ratio is None, then this block shouldn't execute, it means we have a source file with an undetermined aspect ratio so we shouldn't attempt to make it conform to 16:9 -- you guys can definitely come up with a better solution here than mine, but that was the thinking here.

**Comment by yro** [comment]
@mushtaqak This is a lot of really good work! Thank you.

File: video_worker/__init__.py [hunk 14]
```diff
@@ -97,25 +84,24 @@ def test(self):
...
         )
-        if self.setup is True:
+        if self.setup:
             WS.setup = True
 
         WS.run()
         self.settings = WS.settings_dict
 
         if self.encode_profile is None:
-            ErrorObject().print_error(
```
Comment (mushtaqak) [inline video_worker/__init__.py]: @muzaffaryousaf Are we keeping ErrorObject? we removed in previous 2 PRs though.

File: video_worker/tests/test_video_worker.py [hunk 15]
```diff
@@ -0,0 +1,535 @@
...
+        job_id = data.get('jobid', None)
+        expected_workdir = data.get(
+            'workdir',
+            os.path.join(ENCODE_WORK_DIR, job_id) if job_id else ENCODE_WORK_DIR
+        )
+        self.assertEqual(VW.workdir, expected_workdir)
+
+    @data(True, False)
+    @patch('nose.run')
+    def test_video_worker_test(self, nose_result, mock_nose):
```
Comment (mushtaqak) [inline video_worker/tests/test_video_worker.py]: we are working that `test` method has full coverage and is working correctly.

File: video_worker/tests/test_generate_encode.py (hunk 8 above)
Comment (mushtaqak) [inline video_worker/tests/test_generate_encode.py]: I am adding dynamic keys so we don't have to add each time

File: video_worker/tests/test_generate_encode.py (hunk 9 above)
Comment (mushtaqak) [inline video_worker/tests/test_generate_encode.py]: We have tested each other method too.

File: video_worker/tests/test_video_worker.py [hunk 16]
```diff
@@ -0,0 +1,535 @@
...
+        """
+        Test `test` method works correctly.
+        """
+        # Mock nose.run
+        mock_nose.return_value = nose_result
+        result = self.VW.test()
+        self.assertEqual(result, nose_result)
+
+    @data(
```
Comment (mushtaqak) [inline video_worker/tests/test_video_worker.py]: we are working on dynamic key-values so added these intentionally

File: video_worker/tests/test_video_worker.py (hunk 2 above)
Comment (mushtaqak) [inline video_worker/tests/test_video_worker.py]: I couldn't find an alternate way. for each mock to change values dynamically we needed a method and that doesn't take other parameters so I had to chose this approach, happy to listen to other suggestions

File: video_worker/tests/test_video_worker.py [hunk 17]
```diff
@@ -0,0 +1,535 @@
...
+        {
+            'jobid': 'dummy-job-id',
+            'workdir': '',
+            'instance_yaml': 'dummy-instance_yaml.yaml',
+        },
+        {
+            'jobid': 'dummy-job-id',
+            'workdir': '',
+            'instance_yaml': 'dummy-instance_yaml.yaml',
+        },
```
Comment (mushtaqak) [inline video_worker/tests/test_video_worker.py]: My mistake, I will remove

**Comment by muzaffaryousaf** [comment]
@mushtaqak I think @muhammad-ammar 's & @yro 's review is sufficient.

File: video_worker/tests/test_generate_encode.py [hunk 18]
```diff
@@ -0,0 +1,371 @@
...
+                'ffcommand': None,
+                'expected_ffcommand': None
+            }
+        ),
+        # (
+        #     {
+        #         'target_aspect_ratio': False,
+        #         'expected_ffcommand': []
+        #     }
+        # ),
```
Comment (muhammad-ammar) [inline video_worker/tests/test_generate_encode.py]: are we going to remove this?

File: video_worker/tests/test_generate_encode.py [hunk 19]
```diff
@@ -0,0 +1,371 @@
...
+        # (
+        #     {
+        #         'resolution': 480,
+        #         'mezz_resolution': '720x480',
+        #         'mezz_bitrate': 'test-rate',
+        #         'expected_ffcommand': []
+        #     }
+        # )
```
Comment (muhammad-ammar) [inline video_worker/tests/test_generate_encode.py]: So we will add the above commented scenarios when we address the `TODO1`?

File: video_worker/tests/test_generate_encode.py [hunk 20]
```diff
@@ -0,0 +1,371 @@
...
+            'veda_id': None,
+            'expected_ffcommand': ['-movflags', 'faststart']
+        },
+    )
+    @unpack
+    def test_destination(self, file_type, veda_id, expected_ffcommand):
+        """
+        Tests that `_destination` works correctly.
+        """
+        #from nose.tools import set_trace;set_trace()
```
Comment (muhammad-ammar) [inline video_worker/tests/test_generate_encode.py]: left during debugging?

**Comment by mushtaqak** [comment]
Coverage is 62% now - still after some blocks were left to cover :)
```
Name                                Stmts   Miss Branch BrPart  Cover
---------------------------------------------------------------------
video_worker.py                       156      3     54     11    93%
video_worker/abstractions.py          127     81     30      2    31%
video_worker/api_communicate.py       102      3     48      7    93%
video_worker/celeryapp.py              37     16      6      2    53%
video_worker/config.py                 45      1     10      1    96%
video_worker/generate_apitoken.py      31      1      8      1    95%
video_worker/generate_delivery.py      83     53     16      1    31%
video_worker/generate_encode.py       146     25     68     11    79%
video_worker/global_vars.py            18      0      0      0   100%
video_worker/reporting.py              77     46     16      4    42%
video_worker/validate.py               84     66     46      2    15%
video_worker/video_images.py           81     42     16      0    40%
---------------------------------------------------------------------
TOTAL                                 987    337    318     42    62%
----------------------------------------------------------------------
Ran 96 tests in 0.186s
```

**Comment by mushtaqak** [comment]
@muzaffaryousaf COdecov does not include `__init__.py` file in their report so that's why codecov/patch report has this issue. While codecov/project is bound to fail with 85% target.

So, are we good to merge this ?

**Comment by muzaffaryousaf** [comment]
@mushtaqak Yes, we can merge without worrying about 85% target.

**Pull request merged**